File: Trabalho1/heapsort.py
def redoHeap(array, i, size):
    son = int(2*i + 1)
    # print("son: " + repr(array[son]))

    # Checa se o filho da direita existe e se eh maior do que o pai
    if ((son + 1) <= (size - 1)) and ( (array[i] < array[son + 1]) or (array[i] < array[son]) ):

        # Descobre qual dos filhos eh o maior
        if( (array[son] >= array[son+1]) & (array[son] >= array[i]) ):
            pos = son
            array[i], array[son] = array[son], array[i]
            redoHeap(array, pos, size)
        elif ((array[son+1] >= array[son]) & (array[son+1] >= array[i])):
            pos = son + 1
            array[i], array[son+1] = array[son+1], array[i]
            redoHeap(array, pos, size)
        # print("Redoheap2(): " + repr(array))
    # Checa se o filho da esquerda existe e se ele eh maior do que pai
    elif (son <= (size - 1)) and (array[i] < array[son]):
        pos = son
        redoHeap(array, pos, size)
        array[i], array[son] = array[son], array[i]
        # print("Redoheap1(): " + repr(array))

def buildHeap(array, size):
    left = int((size-1)/2)
    while (left >= 0):
        redoHeap(array, left, size)
        left = left - 1

def heapSort(array):
    i = len(array) - 1
    buildHeap(array, len(array))
    while(i > 0):
        array[0], array[i] = array[i], array[0]
        redoHeap(array, 0, i)
        i -= 1

File: Trabalho1/heapSort.c
//HeapSort
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX_SIZE 1000000
//Funcao que realizara a troca de variaveis
void swap(int vetor[], int a, int b){
    int aux;
    aux = vetor[a];
    vetor[a] = vetor[b];
    vetor[b] = aux;
    }

void refaz(int vetor[], int i, int tam){
    //printf("\n\nRefaz...");
    //i eh o indice violado
    int filho = 2*i + 1;
    int aux, pos;
    //O procedimento eh chamado ate que atinja uma folha ou a condicao de heap seja confirmada
    if ( (filho <= tam - 1) || (filho + 1 <= tam - 1) && (vetor[i] < vetor[filho] || vetor[i] < vetor[filho + 1]) ){
        //Checagem da existencia do filho a direita
        if (filho + 1 <= tam - 1 ){
            //Checa se o da esquerda eh o maior
            if ( (vetor[filho] >= vetor[filho + 1]) && (vetor[filho] >= vetor[i]) ){
                pos = filho;
                aux = vetor[i];
                vetor[i] = vetor[filho];
                vetor[filho] = aux;
                refaz(vetor, pos, tam);
            }
            //Filho da direita eh o maior
            else if ( (vetor[filho + 1] >= vetor[filho]) && (vetor[filho + 1] >= vetor[i]) ){
                pos = filho + 1;
                aux = vetor[i];
                vetor[i] = vetor[filho + 1];
                vetor[filho + 1] = aux;
                refaz(vetor, pos, tam);
            }
        }
        //Caso o filho da direita nao exista
        else {
            //Checa se o filho da esquerda eh maior que o pai
            if (vetor[filho] >= vetor[i]) {
                pos = filho;
                refaz(vetor, pos, tam);
                aux = vetor[i];
                vetor[i] = vetor[filho];
                vetor[filho] = aux;
                //i = filho;

            }
        }
    }
}

void constroi (int vetor[], int tam){
    //printf("\nConstroi");
    int esq = (tam - 1)/2 ;
    while (esq >= 0){
        //printf("\nesq = %d", vetor[esq]);
        refaz(vetor, esq, tam);
        esq--;
    }
}

void heapSort(int vetor[], int tam){
    int i = tam-1, aux;
    constroi(vetor, tam);
    while (i > 0){
        //swap(vetor, 0, i);
        aux = vetor[0];
        vetor[0] = vetor[i];
        vetor[i] = aux;
        //Chama o refaz() do comeco do vetor ate o ultimo elemento nao ordenado
        refaz(vetor, 0, i );
        i--;
    }
}

void printArray(int A[], int size){
    int i;
    for (i=0; i < size; i++)
        printf("%d ", A[i]);
    printf("\n");
}

void generateRandomArray(int array[], int n){
	int i;
	for(i = 0; i < n; i++){
		array[i] = rand() % 1001;
	}
}

int *read_ints(int *size, char *input_file){
    //FILE *file = fopen("POPULACAO_INICIAL - FORMATADA.txt", "r");
    FILE *file = fopen(input_file, "r");
    //Conjunto de dados a ser lido
    int *dataset = malloc(MAX_SIZE * sizeof(int));
    int i = 0;

    //Variavel com a quantidade de numeros lidos
    *size = 0;
    //Realiza leitura do arquivo file e armazena o valor na posicao i do dataset
    fscanf(file, "%d", &dataset[i]);

    //feof(filename) eh uma funcao que verifica o fim do arquivo
    while(!feof(file)){
        //printf("%d ", dataset[i]);
        i++;
        fscanf(file, "%d", &dataset[i]);

        (*size)++;
    }
    fclose(file);
    return dataset;
}

void main(){
    clock_t ini_sort, fim_sort;
    double tempoSort, tempoBus;
    int length; 
	//int array[n] = {5, 8, 12, 4, 3, 2, 21, 10, 16, 9};
	int *array = read_ints(&length, "populacao_sem_plano_diretor-2010.txt");
    
    ini_sort = clock();
    heapSort(array, length);
    fim_sort = clock(); 
    
    tempoSort = fim_sort-ini_sort;
    printf("Tempo para ordenacao: %.2f", tempoSort);
}
